src/hostapi/pulseaudio/pa_linux_pulseaudio.c

```diff
@@ -1311,3 +1334,82 @@ double GetStreamCpuLoad( PaStream * s )
 
     return PaUtil_GetCpuLoad( &stream->cpuLoadMeasurer );
 }
+
+/** Extensions */
+static void RenameStreamCb(pa_stream *s, int success, void *userdata)
+{
+    /* Currently does nothing but signal the caller. */
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) userdata;
+    pa_threaded_mainloop_signal( l_ptrStream->mainloop,
+                                 0 );
+}
+
+PaError PaPulseAudio_RenameSource( PaStream *s, const char *streamName )
+{
+    PaPulseAudio_Stream *stream = (PaPulseAudio_Stream *) s;
+    PaError result = paNoError;
+    pa_operation *op = NULL;
+
+    /* Reallocate stream name in memory. */
+    char *newStreamName = (char*)PaUtil_AllocateMemory(strlen(streamName));
+    if ( !newStreamName )
+    {
+        return paInsufficientMemory;
+    }
+    strcpy(newStreamName, streamName);
+
+    PaUtil_FreeMemory( stream->sourceStreamName );
+    stream->sourceStreamName = newStreamName;
+
+    if ( stream->inStream == NULL )
+    {
+        return result;
+    }
+
+    pa_threaded_mainloop_lock( stream->mainloop );
+    op = pa_stream_set_name( stream->inStream, streamName, RenameStreamCb, stream );
+    pa_threaded_mainloop_unlock( stream->mainloop );
+
+    /* Wait for completion. */
+    while (pa_operation_get_state( op ) == PA_OPERATION_RUNNING)
+    {
+        pa_threaded_mainloop_wait( stream->mainloop );
+    }
+
+    return result;
+}
+
+PaError PaPulseAudio_RenameSink( PaStream *s, const char *streamName )
+{
+    PaPulseAudio_Stream *stream = (PaPulseAudio_Stream *) s;
+    PaError result = paNoError;
+    pa_operation *op = NULL;
+    
+    /* Reallocate stream name in memory. */
+    char *newStreamName = (char*)PaUtil_AllocateMemory(strlen(streamName));
+    if ( !newStreamName )
+    {
+        return paInsufficientMemory;
+    }
+    strcpy(newStreamName, streamName);
+
+    PaUtil_FreeMemory( stream->sinkStreamName );
+    stream->sinkStreamName = newStreamName;
+
+    if ( stream->outStream == NULL )
```

Comment:
Actually we shouldn't change name if outStream is NULL. I consider it as error.

Reply:
The use case is to be able to set the name after Pa_OpenStream but prior to Pa_StartStream. The stream would be NULL in that instance, correct?

Reply:
Ok if this a case then 'PaPulseAudio_Initialize'-function should set stream->outStream to NULL (also inStream) as they marked NULL at  OpenStream (https://github.com/illuusio/portaudio/blob/10c8a4050a7186bfcfcd69a3d900b9a62a946f3b/src/hostapi/pulseaudio/pa_linux_pulseaudio.c#L958). Another problem is that you have to call (or if it's possible to make otherwise.. then let me know) OpenStream to initialize PaStream-struct so I don't know how we getting into this situation?

Reply:
I took another look. OpenStream() initially sets both inStream and outStream to NULL. Depending on what's passed into it (inputParameters, outputParameters), one or both of those streams will be created and initialized. Thus, it's probably reasonable to return an error if the corresponding stream is NULL when renaming source or sink. I'll update.